internal/characters/alhaitham/skill.go

```diff
@@ -0,0 +1,259 @@
+
+	c.SetCDWithDelay(action.ActionSkill, 18*60, 15)
+
+	return action.ActionInfo{
+		Frames:          frames.NewAbilFunc(skillTapFrames),
+		AnimationLength: skillTapFrames[action.InvalidAction],
+		CanQueueAfter:   skillTapFrames[action.ActionDash], // earliest cancel
+		State:           action.SkillState,
+	}
+}
+func (c *char) SkillHold() action.ActionInfo {
+	c.Core.Tasks.Add(func() {
+		if c.mirrorCount == 0 { //extra mirror if 0 when cast
+			c.mirrorGain()
+		}
+		c.mirrorGain()
+	}, 23)
+
+	ai := combat.AttackInfo{
+		Abil:               "Universality: An Elaboration on Form (Hold)",
+		ActorIndex:         c.Index,
+		AttackTag:          combat.AttackTagElementalArt,
+		ICDTag:             combat.ICDTagNone,
+		ICDGroup:           combat.ICDGroupDefault,
+		Element:            attributes.Dendro,
+		Durability:         25,
+		Mult:               rushAtk[c.TalentLvlSkill()],
+		FlatDmg:            rushEm[c.TalentLvlSkill()] * c.Stat(attributes.EM),
+		HitlagHaltFrames:   0.09 * 60,
+		HitlagFactor:       0.01,
+		CanBeDefenseHalted: false,
+	}
+	c.Core.QueueAttack(ai, combat.NewCircleHit(c.Core.Combat.Player(), c.Core.Combat.PrimaryTarget(), combat.Point{Y: 2}, 2.25), skillHoldHitmark, skillHoldHitmark)
+
+	c.SetCDWithDelay(action.ActionSkill, 18*60, 23)
+
+	return action.ActionInfo{
+		Frames:          frames.NewAbilFunc(skillHoldFrames),
+		AnimationLength: skillHoldFrames[action.InvalidAction],
+		CanQueueAfter:   skillHoldFrames[action.ActionDash], // earliest cancel
+		State:           action.SkillState,
+	}
+}
+
+func (c *char) mirrorGain() {
+	if c.Base.Cons >= 2 { //triggers on overflow
+		c.c2()
+	}
+	if c.mirrorCount > 2 { //max 3 mirrors at a time.
+		queueOnFrame := false              //var tracks if this is the first overflowing in this frame
+		if c.Core.F == c.lastInfusionSrc { //check if c.lastinfusion has already been called on this frame
+			queueOnFrame = true
+		}
+		if !queueOnFrame { //this avoids multiple queues of mirror loss if mirror overflow multiple times in same frame
+			c.lastInfusionSrc = c.Core.F
+			c.Core.Tasks.Add(c.mirrorLoss(c.Core.F), 234)
+		}
+		c.Core.Log.NewEvent("mirror overflowed", glog.LogCharacterEvent, c.Index)
+
+		if c.Base.Cons >= 6 {
+			c.c6()
+		}
+		return
+	}
+	if c.mirrorCount == 0 {
+		c.lastInfusionSrc = c.Core.F
+		c.Core.Tasks.Add(c.mirrorLoss(c.Core.F), 234)
+		c.Core.Log.NewEvent("infusion added", glog.LogCharacterEvent, c.Index)
+
+	}
+	c.mirrorCount++
+	c.recentlyMirrorGain = true
+	c.Core.Log.NewEvent("Gained 1 mirror", glog.LogCharacterEvent, c.Index)
+
+}
+
+func (c *char) mirrorLoss(src int) func() {
+	return func() {
+		if c.lastInfusionSrc != src {
+			c.Core.Log.NewEvent("mirror decrease ignored, src diff", glog.LogCharacterEvent, c.Index).
+				Write("src", src).
+				Write("new src", c.lastInfusionSrc)
+			return
+		}
+		if c.mirrorCount == 0 { //for case when you swap out and have mirrorloss queue'd. T
+			//TODO:change the lastinfusionsrc on swap event perhaps?
```

Comment:
Please set c.lastInfusionSrc to something like -1 in onExitField and remove the TODO.